### PR TITLE
tpm2_loadexternal: Fix issue where a restricted attr couldn't be set

### LINKED ISSUE #2779: Tool to go from PEM of pulic key to TPM2B_PUBLIC would be nice
Google's cloud environment provides a REST API that returns the public key of the cloud instance's EK as a PEM blob.  When I try to `tpm2 loadexternal` it to eventually get the EKpub as a `TPM2B_PUBLIC`, it fails every time as long as I have the `restricted` attribute.  I don't understand why that would be:

```
$ tpm2 loadexternal -L nullpolicy -Grsa -a 'fixedtpm|fixedparent|sensitivedataorigin|adminwithpolicy|restricted|decrypt' -C n -u nek.pem -c nek-wrong.ctx
WARNING:esys:src/tss2-esys/api/Esys_LoadExternal.c:314:Esys_LoadExternal_Finish() Received TPM Error
ERROR:esys:src/tss2-esys/api/Esys_LoadExternal.c:108:Esys_LoadExternal() Esys Finish ErrorCode (0x000002d6)
ERROR: Esys_LoadExternal(0x2D6) - tpm:parameter(2):unsupported symmetric algorithm or key size or not appropriate for instance
ERROR: Unable to run loadexternal
$ 
```

```
$ # This works, but the name is wrong because restricted is missing:
$ tpm2 loadexternal -L nullpolicy -Grsa -a 'fixedtpm|fixedparent|sensitivedataorigin|adminwithpolicy|decrypt' -C n -u nek.pem -c nek-wrong.ctx
name: 000bf3b2e878796925c719c050497c8eca618a32e74ba2d287af6af718e72fa24113
```

I've tried using `-G rsa2048:oeap:aes256cfb` to get around `unsupported symmetric algorithm or key size or not appropriate for instance`, but no dice:

```
$ tpm2 loadexternal -L nullpolicy -G rsa2048:oeap:aes256cfb -a 'fixedtpm|fixedparent|sensitivedataorigin|adminwithpolicy|restricted|decrypt' -C n -u nek.pem -c nek-wrong.ctx
ERROR: Unsupported key type, got: "rsa2048:oeap:aes256cfb"
ERROR: Unable to run loadexternal
$ 
```

And no variation of that succeeds.

### PR DESCRIPTION
**Pull request by idesai**
Fixes #2779

This should now work
loadexternal -G rsa2048:aes128cfb -C n -u public.pem -c key.ctx \
-a 'fixedtpm|fixedparent|sensitivedataorigin|adminwithpolicy|restricted|decrypt'

Signed-off-by: Imran Desai <imran.desai@intel.com>

### CODE REVIEW

File: tools/tpm2_loadexternal.c [hunk 1]
```diff
@@ -187,6 +176,39 @@ static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {
         },
     };
 
+
+    TPMI_ALG_PUBLIC alg = TPM2_ALG_NULL;
+    bool is_ext_param_specified = false;
+    if (ctx.key_type) {
+        alg = tpm2_alg_util_from_optarg(ctx.key_type,
```
Comment (idesai) [inline tools/tpm2_loadexternal.c:183]: this call is not required

File: tools/tpm2_loadexternal.c [hunk 2]
```diff
@@ -187,6 +176,39 @@ static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {
...
+
+    if (ctx.public_key_path) {
+        bool result = tpm2_openssl_load_public(ctx.public_key_path, alg, &pub);
+        if (!result) {
+            return tool_rc_general_error;
+        }
+    }
+
+    if (is_ext_param_specified) {
+        bool result = tpm2_alg_util_handle_ext_alg(ctx.key_type, &pub);
```
Comment (idesai) [inline tools/tpm2_loadexternal.c:205]: read the params in the first call and then re-assign after public key is loaded

**Review by idesai — COMMENTED** [review]
Although, this change fixes the immediate issue at hand, has caused some integration tests to fail. Analyze and fix accordingly.

**Comment by nicowilliams** [comment]
I think I'm running this code, but I'm getting:

```
$ tpm2 loadexternal -G rsa2048:aes128cfb -C n -u ek.pem -c EK.ctx -a 'fixedtpm|fixedparent|sensitivedataorigin|adminwithpolicy|restricted|decypt'
ERROR: Unsupported key type, got: "rsa2048:aes128cfb"
ERROR: Unable to run loadexternal
```

**Comment by williamcroberts** [comment]
> I think I'm running this code, but I'm getting:
> 
> ```
> $ tpm2 loadexternal -G rsa2048:aes128cfb -C n -u ek.pem -c EK.ctx -a 'fixedtpm|fixedparent|sensitivedataorigin|adminwithpolicy|restricted|decypt'
> ERROR: Unsupported key type, got: "rsa2048:aes128cfb"
> ERROR: Unable to run loadexternal
> ```
> 
> I can't seem to find an alg name that would work.

Try:
rsa2048:null:aes128cfb

**Comment by nicowilliams** [comment]
Oh, no, I'm not running these bits.  Sorry, let me make sure I am first.

**Comment by nicowilliams** [comment]
OK, I can load it now.  Is there a way I can print it its `TPM2B_PUBLIC`?

My goal is to take a public key in PEM format from the Google Shielded VM REST API and output an EK.pub that I can use with `tpm2 duplicate` and `tpm2 makecredential` (which I use with `--tcti=none`), or, failing that, that I could use the PEM as the EK for `tpm2 duplicate` and `tpm2 makecredential`.

**Comment by nicowilliams** [comment]
Ahhh, got it.  `tpm2 readpublic -o EK.pub -c EK.ctx`.  I also had to set the correct policy when using `tpm2 loadexternal`, and then the cryptographic name came out matching.  Thanks!

**Pull request merged**